tests/common/test_io.py

```diff
@@ -87,3 +92,40 @@ def test_attach_stderr_handler():
     assert c.stdout == ""
     assert "test message" in c.stderr
     assert debug_message in c.stderr
+
+
+@pytest.mark.parametrize(
+    "thread_class,should_fail",
+    [
+        (ThreadLimitedThreadPoolExecutor, False),
+        (ThreadPoolExecutor, True),
+    ],
+)
+def test_thread_limited_executor_handles_thread_limit(
+    thread_class, should_fail, mocker
+):
+    """
+    Ensure our custom ThreadLimitedThreadPoolExecutor class does what it is
+    intended to do, which is not raise a RuntimeError if the max workers limit
+    is reached.
+
+    This only affects users of particular systems like HPC clusters.
+
+    Some historical background can be found here:
+        - https://github.com/conda/conda/pull/6653
+        - https://github.com/conda/conda/issues/7040
+    """
+    jobs = 10
+
+    if should_fail:
+        mocker.patch(
+            "concurrent.futures.ThreadPoolExecutor._adjust_thread_count",
+            side_effect=RuntimeError,
+        )
+
+    with thread_class(max_workers=2) as executor:
+        if should_fail:
+            with pytest.raises(RuntimeError):
+                _ = [executor.submit(time.sleep, 0.1) for _ in range(jobs)]
+        else:
+            _ = [executor.submit(time.sleep, 0.1) for _ in range(jobs)]
```

Comment:
We use `contextlib.nullcontext` to avoid duplicate code:

```suggestion
        with pytest.raises(RuntimeError) if should_fail else nullcontext():
            assert [executor.submit(time.sleep, 0.1) for _ in range(10)]
```